crates/iceberg-catalog/src/service/mod.rs

```diff
@@ -131,10 +139,55 @@ impl From<ProjectId> for uuid::Uuid {
 impl ProjectId {
     #[must_use]
     pub fn new(id: uuid::Uuid) -> Self {
+        Self(id.to_string())
+    }
+
+    /// Create a new project id from a string.
+    ///
+    /// # Errors
+    /// Returns an error if the provided string is not a valid project id.
+    /// Valid project ids may only contain alphanumeric characters, hyphens and underscores.
+    pub fn try_new(id: String) -> Result<Self> {
+        // Only allow the following characters in the project id:
+        // - Alphanumeric characters
+        // - Hyphens
+        // - Underscores
+
+        if id
+            .chars()
+            .all(|c| c.is_alphanumeric() || c == '-' || c == '_')
+        {
+            Ok(Self(id))
+        } else {
+            Err(ErrorModel::builder()
+                .code(StatusCode::BAD_REQUEST.into())
+                .message(format!(
+                    "Project IDs may only contain alphanumeric characters, hyphens and underscores. Got: `{id}`",
+                ))
+                .r#type("MalformedProjectID".to_string())
+                .build()
+                .into())
```

Comment:
Let's use the variant constructor, eventually, I'd like us to get rid of the builder altogether